openfe/tests/protocols/openmm_septop/test_septop_protocol.py

```diff
@@ -0,0 +1,1261 @@
+        )
+
+
+def test_create_default_protocol(default_settings):
+    # this is roughly how it should be created
+    protocol = SepTopProtocol(
+        settings=default_settings,
+    )
+    assert protocol
+
+
+def test_serialize_protocol(default_settings):
+    protocol = SepTopProtocol(
+        settings=default_settings,
+    )
+
+    ser = protocol.to_dict()
+    ret = SepTopProtocol.from_dict(ser)
+    assert protocol == ret
+
+
+def test_create_independent_repeat_ids(
+    benzene_complex_system, toluene_complex_system, default_settings
+):
+    # if we create two dags each with 3 repeats, they should give 6 repeat_ids
+    # this allows multiple DAGs in flight for one Transformation that don't clash on gather
+    # Default protocol is 1 repeat, change to 3 repeats
+    default_settings.protocol_repeats = 3
+    protocol = SepTopProtocol(
+            settings=default_settings,
+    )
+
+    dag1 = protocol.create(
+        stateA=benzene_complex_system,
+        stateB=toluene_complex_system,
+        mapping=None,
+    )
+    dag2 = protocol.create(
+        stateA=benzene_complex_system,
+        stateB=toluene_complex_system,
+        mapping=None,
+    )
+    # print([u for u in dag1.protocol_units])
+    repeat_ids = set()
+    for u in dag1.protocol_units:
+        repeat_ids.add(u.inputs['repeat_id'])
+    for u in dag2.protocol_units:
+        repeat_ids.add(u.inputs['repeat_id'])
+
+    # There are 4 units per repeat per DAG: 4 * 3 * 2 = 24
+    assert len(repeat_ids) == 24
+
+
+def test_check_alchem_charge_diff(charged_benzene_modifications):
+    errmsg = "A charge difference of 1"
+    with pytest.raises(ValueError, match=errmsg):
+        _check_alchemical_charge_difference(
+            charged_benzene_modifications["benzene"],
+            charged_benzene_modifications["benzoic_acid"],
+        )
+
+
+def test_charge_error_create(
+    charged_benzene_modifications, T4_protein_component, default_settings
+):
+    # if we create two dags each with 3 repeats, they should give 6 repeat_ids
+    # this allows multiple DAGs in flight for one Transformation that don't clash on gather
+    # Default protocol is 1 repeat, change to 3 repeats
+    default_settings.protocol_repeats = 3
+    protocol = SepTopProtocol(
+            settings=default_settings,
+    )
+    stateA = ChemicalSystem({
+        'benzene': charged_benzene_modifications['benzene'],
+        'protein': T4_protein_component,
+        'solvent': SolventComponent()
+    })
+
+    stateB = ChemicalSystem({
+        'benzoic': charged_benzene_modifications['benzoic_acid'],
+        'protein': T4_protein_component,
+        'solvent': SolventComponent(),
+    })
+    errmsg = "A charge difference of 1"
+    with pytest.raises(ValueError, match=errmsg):
+        protocol.create(
+            stateA=stateA,
+            stateB=stateB,
+            mapping=None,
+        )
+
+
+def test_validate_complex_endstates_protcomp_stateA(
```

Comment:
Not sure what the preferred pattern is, but for these pairs of tests that test validation in state A and B it would reduce the amount of repeated code if we could use pytest parameterized on them.